[PEAUTY-135] payment order logic first development

harvartz opened the pull request:
## 📄 [PF-135] payment order logic first development
Jira : [PEAUTY-135](https://multicampusuplus.atlassian.net/browse/PEAUTY-135?atlOrigin=eyJpIjoiN2E0ODVhYjI2MjhjNGY4OGI2ZWQ4YWRmMDM0Y2VlMDkiLCJwIjoiaiJ9)

## ✨ 변경 사항
- [x] 기능 추가/변경 설명
- [x] 버그 수정 설명
- [x] 문서 수정 설명

<!-- Pull Request의 설명을 추가하세요. -->
-  사용자가 주문한 결제에 대한 정보를 서버로 저장하는 API입니다.
- 추후 포트원에서 받아온 결제 데이터와 서버에 저장한 데이터와 유효한지 검증하기 위해 만들었습니다.
- customer build gradle쪽에 payment 모듈에 대한 의존성을 추가했습니다.

## 📅 작업 일정
<!-- 해당 작업을 수행하는데 예상했던 공수와 실제 소요되었던 공수를 기입해주세요. -->
- Expected MD: 2MD
- Actual MD: 2MD
### Difference reason (If correct, no need.)
- None.

## ✔️ 확인 사항

- [x] 코드가 잘 작동하는지 확인했나요?
- [x] 새로운 기능에 대한 테스트가 추가되었나요?
- [x] 문서가 업데이트되었나요?

harvartz committed:
feat: 도메인, JPA 영속 객체 초기 설계

harvartz committed:
Merge branch 'release/v0.1.0' of https://github.com/peauty/peauty-BE into feature/PEAUTY-135

harvartz committed:
Merge branch 'release/v0.1.0' of https://github.com/peauty/peauty-BE into feature/PEAUTY-135

harvartz committed:
feat: 결제 모듈 환경 구축
- 공통 Response, Exception 생성
- build.gradle 설정 변경

harvartz committed:
feat: 결제 도메인 설계

harvartz committed:
feat: 주문 로직 1차 초기 구현

harvartz committed:
feat: 결제 API 환경 구축 (20%)

harvartz committed:
Merge branch 'release/v0.1.0' of https://github.com/peauty/peauty-BE into feature/PEAUTY-135

weejinyoung commented on peauty-domain/src/main/java/com/peauty/domain/payment/Order.java:
```diff
@@ -0,0 +1,43 @@
+    private Long threadId;
+    private Integer cost;
+    private Boolean paymentStatus;
+    private LocalDateTime orderDate;
```
boolean 타입으로 할 거면 
as is: paymentStatus
to be: isPaymentCompleted
추천드립니다

weejinyoung commented on peauty-payment/src/main/java/com/peauty/payment/business/PaymentPort.java:
```diff
@@ -0,0 +1,7 @@
+
+import com.peauty.domain.payment.Order;
+
+public interface PaymentPort {
+    Order saveNewOrder(Order order);
+}
```
save 로 할 것인가 saveNewOrder 로 할 것인가?!

harvartz replied:
다른 곳에서는 사용하지 않을 것 같아서 save로 하겠습니다 : )

weejinyoung commented on peauty-payment/src/main/java/com/peauty/payment/business/PaymentServiceImpl.java:
```diff
@@ -0,0 +1,59 @@
+    public OrderResult saveOrder(Long userId, Long processId, Long threadId, OrderCommand command) {
+        // TODO: 유저 검증
+        // TODO: 요청서 & 견적서 정보 검증 (요청서, 견적서가 DB에 있는지 검증)
+
+        Order orderToSave = command.toDomain(userId, processId, threadId);
+        UUID uuid = UUID.randomUUID();
+        orderToSave.updateOrderUuid(uuid.toString());
+        orderToSave.updateOrderDate(LocalDateTime.now());
+        // TODO: status 어떻게 관리할지 확인하기!
+        orderToSave.updatePaymentStatus(false);
+
```
toDomain 인자에 orderUuid, orderDate, paymentStatus 가 추가되면 좋지 않을까요
그것이 데이터 정합성이 깨지는 것을 방지할 수 있는 좋은 방법이지 않을까 싶어요

harvartz replied:
이런 추가적으로 나오는 데이터에 대해서 하나의 DTO로 담아서 보내주는 것과 따로 보내 주는 것에 대해서 고민이 많았는데 이런 관점이 있는지 생각을 못했습니다!

파라미터만으로 어떤 데이터가 들어갔는지 안들어갔는지 확인하면서 데이터 정합성을 지킬 수 있다는 좋은 관점을 배우고 갑니다!

Shanate commented on peauty-domain/src/main/java/com/peauty/domain/payment/Order.java:
```diff
@@ -0,0 +1,43 @@
+    public void transferReservationCost(Integer cost) {
+        Double halfAmount = cost * 0.5;
+        // this.cost = (Integer) (Math.round(halfAmount/ 100.0) * 100);
+        this.cost = Math.toIntExact(Math.round(halfAmount / 100.0) * 100);
```
백 아래 단위는 사용하지 않기 위해 이 수학적 장치를 쓴 것인가요?

harvartz replied:
넵 맞습니다. 결제하는데 10단위까지 내려오면 그렇게 좋아보이지 않을 것 같아서 만들어봤습니다!

Shanate commented on peauty-domain/src/main/java/com/peauty/domain/payment/Order.java:
좋은 도메인인거 같습니다

Shanate commented on peauty-payment/src/main/java/com/peauty/payment/presentation/PaymentController.java:
```diff
@@ -0,0 +1,50 @@
+    // 주문 정보 API
+    // TODO: 이런 URL을 가져가야 하는 것이 맞는가?
+    // URL은 받되 유효성 검증은 좀 나중에 진행하자.
+    @PostMapping("/users/{userId}/processes/{processId}/threads/{threadId}/order")
```
적절한거같습니다

Shanate reviewed (approved):
결제가 완성됐을 때에 대한 기대가 많이 크네요
LGTM

weejinyoung commented on peauty-payment/src/main/java/com/peauty/payment/presentation/PeautyPaymentExceptionHandler.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.ArrayList;
+import java.util.List;
+
+@Slf4j
+@RestControllerAdvice
+public class PeautyPaymentExceptionHandler {
+
```
요거 이제 customer api 에 기생하는 라이브러리가 된 이상 필요 없을 것 같아욤

weejinyoung commented on peauty-payment/src/main/java/com/peauty/payment/presentation/PeautyPaymentResponseAdvice.java:
```diff
@@ -0,0 +1,61 @@
+import org.springframework.http.server.ServerHttpResponse;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice;
+
+@ControllerAdvice(
+        value = {
+                "com.peauty.payment.presentation"
+        }
+)
+public class PeautyPaymentResponseAdvice implements ResponseBodyAdvice<Object> {
```
ExceptionHandler 는 없어도 될 것 같지만... 이거는 
```
@ControllerAdvice(
        value = {
                "com.peauty.payment.presentation"
        }
)
```
이렇게 패키지 스캔을 하는 것 같아서 일단 있어야 할 것 같아요

harvartz replied:
이 멀티모듈은 알 것 같으면서도 생각보다 복잡하네요ㅠ 다음에 여유있을때 학습하면서 내부 구조에 대해서 공부해봐야 겠습니다ㅠ

weejinyoung commented on peauty-persistence/src/main/java/com/peauty/persistence/payment/OrderEntity.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Column(name = "cost", nullable = false)
+    private Integer cost;
+
+    @Column(name = "payment_status", nullable = false)
+    private Boolean paymentStatus;
+
```
이것도 is~ 로 바꿔주시면 감사하겠습니다

weejinyoung reviewed (changes requested):
고생하셨습니다 몇 가지 리뷰 남겨요!

harvartz committed:
feat: 리뷰 반영
- isPaymentCompleted -> 네이밍에 조금 더 신경쓰자.
- saveNewOrder -> 네이밍에 더 신경쓰자
- 추가되는 파라미터를 직접 명시하여 정합성을 지키는데 힘쓰자.
